### PR TITLE
feat(cloud-function): implement Cloud Function + deploy to stage/prod

### PR DESCRIPTION
**Pull request by caugner**
## Summary

Relates to MP-277 and MP-278.

### Problem

We're migrating to GCP.

In AWS, we have been using CloudFront behaviors and Lambdas to route requests to our API and S3 and set cache policies, but GCP has no equivalent services.

### Solution

Implement a Cloud Function that routes requests to our API and Cloud Storage, setting `Cache-Control` headers where applicable.

---

## How did you test this change?

Initially, we deployed the Cloud Function to a new `xyz` environment at https://developer.allizom.xyz, and now it is also deployed to the `stage` environment at https://developer.allizom.org/.

### CODE REVIEW

File: gcp/function/src/handlers/kevel.ts [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    } catch (e) {
+      console.error(e);
+      return res.status(500).end();
+    }
+  } else if (pathname === "/pong/viewed") {
+    if (req.method !== "POST") {
+      return {
+        status: 405,
+        statusDescription: "METHOD_NOT_ALLOWED",
+      };
```
Comment (caugner) [inline gcp/function/src/handlers/kevel.ts:172]:
```suggestion
      return res.status(405).end();
```

File: gcp/function/src/app.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import { redirectFundamental } from "./middlewares/redirectFundamental.js";
+import { redirectLocale } from "./middlewares/redirectLocale.js";
+import { redirectTrailingSlash } from "./middlewares/redirectTrailingSlash.js";
+
+const mainRouter = Router();
+const proxyContent = createContentProxy();
+mainRouter.use(redirectLeadingSlash);
+mainRouter.get("/bcd/api/*", proxyBcdApi());
+mainRouter.all("/api/v1/stripe/plans", plans);
+mainRouter.all("/plus/plans.json", plans);
```
Comment (caugner) [inline gcp/function/src/app.ts:23]:
```suggestion
```

File: gcp.sh [hunk 3]
```diff
@@ -0,0 +1 @@
+cd gcp/function && npm i && npm run build-deploy-clean
```
Comment (caugner) [inline gcp.sh:1]: remove

File: libs/constants/index.js [hunk 4]
```diff
@@ -106,13 +106,21 @@ export const CSP_DIRECTIVES = {
...
     "interactive-examples.prod.mdn.mozilla.net",
+    "yari-demos.prod.mdn.mozit.cloud",
+
+    // Stage.
     "interactive-examples.mdn.allizom.net",
+    "interactive-examples.mdn.allizom.net",
+    "yari-demos.stage.mdn.mozit.cloud",
+
+    // XYZ.
+    "interactive-examples.developer.allizom.xyz",
```
Comment (caugner) [inline libs/constants/index.js:120]:
```suggestion
```

File: libs/locale-utils/index.js [hunk 5]
```diff
@@ -12,27 +12,51 @@ const VALID_LOCALES_LIST = [...VALID_LOCALES.values()];
 // From https://github.com/aws-samples/cloudfront-authorization-at-edge/blob/01c1bc843d478977005bde86f5834ce76c479eec/src/lambda-edge/shared/shared.ts#L216
 // but rewritten in JavaScript (from TypeScript).
 function extractCookiesFromHeaders(headers) {
+  let value = headers["cookie"];
```
Comment (caugner) [inline libs/locale-utils/index.js:15]: @fiji-flo The changes in this file need review.

File: .github/workflows/xyz-build.yml [hunk 6]
```diff
@@ -0,0 +1,206 @@
+name: XYZ Build (GCP)
+
+on:
+  push:
+    branches:
+      - gcp
```
Comment (caugner) [inline .github/workflows/xyz-build.yml:6]:
```suggestion
      - gcp
      - main
```

Comment (caugner) [reply .github/workflows/xyz-build.yml:6]: Added `on: workflow_dispatch` instead.

File: .github/workflows/stage-build.yml [hunk 7]
```diff
@@ -18,6 +18,12 @@ on:
   schedule:
     # * is a special character in YAML so you have to quote this string
     - cron: "0 */24 * * *"
+  push:
+    branches:
+      - gcp
+    paths:
+      - .github/workflows/stage-build.yml
+      - gcp/function/**
```
Comment (caugner) [inline .github/workflows/stage-build.yml:26]: ⚠️ Revert before merge.

File: .github/workflows/prod-build.yml [hunk 8]
```diff
@@ -263,18 +286,105 @@ jobs:
...
+        run: |-
+          gsutil -m -h "Cache-Control:public, max-age=86400" rsync -crj html,json,txt client/build gs://content-prod-mdn/main
+
+      # Deploy Function
+      - name: Authenticate with GCP
+        if: ${{ ! vars.SKIP_FUNCTION || !vars.SKIP_INVALIDATE }}
+        uses: google-github-actions/auth@v1
+        with:
+          token_format: access_token
+          service_account: deploy-prod-mdn-ingre@${{ secrets.GCP_PROJECT_NAME }}.iam.gserviceaccount.com
```
[the remark this answers is not in the document]
Comment (bkochendorfer) [reply .github/workflows/prod-build.yml:326]: We get some more letters!
```suggestion
          service_account: deploy-prod-prod-mdn-ingress@${{ secrets.GCP_PROJECT_NAME }}.iam.gserviceaccount.com
```

File: .github/workflows/prod-build.yml [hunk 9]
```diff
@@ -263,18 +286,105 @@ jobs:
...
+          CONTENT_TRANSLATED_ROOT: ${{ github.workspace }}/mdn/translated-content/files
+        run: |
+          npm ci
+          npm run redirects
+
+      - name: Deploy Function
+        if: ${{ ! vars.SKIP_FUNCTION }}
+        run: |-
+          for region in europe-west1 us-west1 asia-east1; do
+            gcloud beta functions deploy mdn-prod-$region \
```
[the remark this answers is not in the document]
Comment (bkochendorfer) [reply .github/workflows/prod-build.yml:349]:
```suggestion
            gcloud beta functions deploy mdn-prod-prod-$region \
```

File: gcp/function/src/handlers/content.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    xfwd: true,
+    selfHandleResponse: true,
+    onProxyReq: fixRequestBody,
+    onProxyRes: responseInterceptor(
+      async (responseBuffer, proxyRes, req, res) => {
+        withContentResponseHeaders(proxyRes, req, res);
+        if (proxyRes.statusCode === 404) {
+          const response = await fetch(`${target}${NOT_FOUND_PATH}`);
+          res.setHeader("Content-Type", "text/html");
+          return Buffer.from(await response.arrayBuffer());
```
Comment (fiji-flo) [inline gcp/function/src/handlers/content.ts:29]: We could cache this with a global `let 404_BUFFER;`

File: gcp/function/src/middlewares/redirectTrailingSlash.ts [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+//       for "/{locale}/account/?" and "/account/?" that redirect to
+//       "/{locale}/settings" and "/settings" respectively. The other
+//       cases can be either redirected or deleted eventually as well.
+//       The goal is to eventually remove the code that uses
+//       LEGACY_URI_NEEDING_TRAILING_SLASH.
+const LEGACY_URI_NEEDING_TRAILING_SLASH = new RegExp(
+  `^(?:${[...LOCALE_URI_WITHOUT_TRAILING_SLASH].join(
+    "|"
+  )})?/(?:account|contribute|maintenance-mode|payments)/?$`
+);
```
Comment (fiji-flo) [inline gcp/function/src/middlewares/redirectTrailingSlash.ts:26]:
```suggestion
```

File: gcp/function/src/middlewares/redirectTrailingSlash.ts [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+    // from S3, we need to strip the trailing slash before it reaches
+    // S3. This is required because we store the home pages in S3 as
+    // their path name itself, for example "en-us" for the English home
+    // page, not "en-us/index.html", which is what S3 would look for if
+    // we left the trailing slash.
+    requestURI = requestURI.slice(0, -1);
+  } else if (
+    requestURI.endsWith("/") &&
+    !LEGACY_URI_NEEDING_TRAILING_SLASH.test(requestURILowerCase)
+  ) {
```
Comment (fiji-flo) [inline gcp/function/src/middlewares/redirectTrailingSlash.ts:61]:
```suggestion
  } else if (requestURI.endsWith("/")) {
```

File: gcp/function/src/middlewares/redirectTrailingSlash.ts (hunk 11 above)
Comment (fiji-flo) [reply gcp/function/src/middlewares/redirectTrailingSlash.ts:26]: not needed, `/accounts/` is handled in fundamental all others seem to be dependent on kuma.

File: libs/pong/coding.d.ts [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+   * @param {string} signSecret - The signing secret.
+   */
+  constructor(signSecret: string);
+  /**
+   * The signing secret.
+   * @type {string}
+   */
+  signSecret: string;
+  encodeAndSign(s?: string): string;
+  decodeAndVerify(tuple?: string): string;
```
Comment (fiji-flo) [inline libs/pong/coding.d.ts:13]:
```suggestion
  decodeAndVerify(tuple?: string): string | null;
```

**Review by fiji-flo — APPROVED** [review]
Great job!

File: gcp/function/package.json [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+  "author": "",
+  "type": "module",
+  "main": "src/index.js",
+  "scripts": {
+    "build": "tsc -b",
+    "copy-internal": "rm -rf ./src/internal && cp -R ../../libs ./src/internal",
+    "gcp-build": "npm run build",
+    "prepare": "([ ! -e ../../libs ] || npm run copy-internal)",
+    "proxy": "ts-node src/proxy.ts",
+    "redirects": "ts-node src/build.ts",
```
[the remark this answers is not in the document]
Comment (caugner) [reply gcp/function/package.json]: Fixed in 4ac61b8fb.

File: gcp/function/src/handlers/content.ts (hunk 10 above)
Comment (caugner) [reply gcp/function/src/handlers/content.ts]: Fixed in 42705d60c, works like a charm.

File: cloud-function/src/handlers/proxy-content.ts [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+  onProxyRes: responseInterceptor(
+    async (responseBuffer, proxyRes, req, res) => {
+      withContentResponseHeaders(proxyRes, req, res);
+      if (proxyRes.statusCode === 404) {
+        if (!notFoundBuffer) {
+          const response = await fetch(`${target}${NOT_FOUND_PATH}`);
+          notFoundBuffer = await response.arrayBuffer();
+        }
+        res.setHeader("Content-Type", "text/html");
+        return Buffer.from(notFoundBuffer);
```
Comment (fiji-flo) [inline cloud-function/src/handlers/proxy-content.ts:34]:
```suggestion
        const tryHtml = await fetch(`${target}${req.url}/index.html`);
        if (tryHtml.ok) {
          res.statusCode = 200;
          return Buffer.from(await tryHtml.arrayBuffer());
        } else if (!notFoundBuffer) {
          const response = await fetch(`${target}${NOT_FOUND_PATH}`);
          notFoundBuffer = await response.arrayBuffer();
        }
        res.setHeader("Content-Type", "text/html");
        return Buffer.from(notFoundBuffer);
```

**Review by fiji-flo — APPROVED** [review]
Looking great

**Pull request merged**